test/integration/resourcemanager/csrapprover/csrapprover_test.go

```diff
@@ -265,11 +281,265 @@ var _ = Describe("Kubelet Server CertificateSigningRequest Approver Controller t
+							HaveField("Type", certificatesv1.CertificateApproved),
+							HaveField("Reason", "RequestApproved"),
+							HaveField("Message", "Approving gardener-node-agent certificate CSR (all checks passed)"),
+						)))
+					}).Should(Succeed())
+				})
+			})
+
+			Context("constraints violated", func() {
+				runTest := func(expectedReason string, argPtrs ...*string) {
+					It("should deny the CSR", func() {
+						var args []interface{}
+						for _, arg := range argPtrs {
+							args = append(args, *arg)
+						}
+						EventuallyWithOffset(1, func(g Gomega) {
+							g.Expect(testClientNodeAgent.Get(ctx, client.ObjectKeyFromObject(csr), csr)).To(Succeed())
+							g.Expect(csr.Status.Conditions).To(ContainElement(And(
+								HaveField("Type", certificatesv1.CertificateDenied),
+								HaveField("Reason", "RequestDenied"),
+								HaveField("Message", And(
+									ContainSubstring("Denying gardener-node-agent certificate CSR"),
+									ContainSubstring(expectedReason, args...),
+								)),
+							)))
+						}).Should(Succeed())
+					})
+				}
+
+				Context("no machine", func() {
+					runTest("machine %q does not exist", &machineName)
+				})
+
+				Context("a certificate for a different machine is requested", func() {
+					var otherMachineName string
+
+					BeforeEach(func() {
+						machine2 := machine.DeepCopy()
+						machine2.Name = "foo-bar-machine"
+						otherMachineName = "gardener.cloud:node-agent:machine:" + machine2.Name
+						certificateSubject = &pkix.Name{
+							CommonName: otherMachineName,
+						}
+						createMachine(machine, false)
+						createMachine(machine2, false)
+					})
+
+					runTest("username %q and commonName %q do not match", &userNameNodeAgent, &otherMachineName)
+				})
+			})
+		})
+
+		Context("bootstrap user", func() {
+			JustBeforeEach(func() {
+				By("Create CertificateSigningRequest")
+				Expect(testClientBootstrap.Create(ctx, csr)).To(Succeed())
+				log.Info("Created CertificateSigningRequest for test", "certificateSigningRequest", client.ObjectKeyFromObject(csr))
+
+				DeferCleanup(func() {
+					By("Delete CertificateSigningRequest")
+					Expect(client.IgnoreNotFound(testClientBootstrap.Delete(ctx, csr))).To(Succeed())
+				})
+			})
+
+			Context("constraints fulfilled", func() {
+				BeforeEach(func() {
+					createMachine(machine, false)
+				})
+
+				It("should approve the CSR", func() {
+					Eventually(func(g Gomega) {
+						g.Expect(testClientBootstrap.Get(ctx, client.ObjectKeyFromObject(csr), csr)).To(Succeed())
+						g.Expect(csr.Status.Conditions).To(ContainElement(And(
+							HaveField("Type", certificatesv1.CertificateApproved),
+							HaveField("Reason", "RequestApproved"),
+							HaveField("Message", "Approving gardener-node-agent certificate CSR (all checks passed)"),
+						)))
+					}).Should(Succeed())
+				})
+			})
+
+			Context("constraints violated", func() {
+				runTest := func(expectedReason string, argPtrs ...*string) {
```

Comment:
The `runTest` function seems repeated across contexts with the only difference being the client. Is it not possible to reuse it?